OCS endpoint to list remote shares

rullzer opened the pull request:
Required for #9098
- Move OCS endpoint `../remote_shares` -> `../pending_shares`
- Introduce new endpoint `../remote_shares` that allows listing of incoming remote shares.

@nickvergessen since you wrote the original `../remote_shares` stuff are you fine with this?

I'm not really happy with this. But as stated in https://github.com/owncloud/core/issues/9098#issuecomment-143311672 this is the cleanest hack I can think of. Comments?

CC: @cmonteroluque @DeepDiver1975

rullzer committed:
Move OCS endpoint ../remote_share to ../pending_shares

rullzer committed:
Added '../remote_shares' endpoint

* list incoming remote shares at 'remote_shares'
* get per share info at 'remote_shares/<ID>'
* delete remote share with a DELETE to 'remote_shares/<ID>'

phil-davis commented on apps/files_sharing/api/remote.php:
```diff
@@ -90,4 +90,79 @@ public static function declineShare($params) {
+		$shareInfo = $externalManager->getShare($params['id']);
+
+		if ($shareInfo === false) {
+			return new \OC_OCS_Result(null, 404, 'share does not exists');
```
s/exists/exist/

phil-davis commented on apps/files_sharing/api/remote.php:
```diff
@@ -90,4 +90,79 @@ public static function declineShare($params) {
+		$shareInfo = $externalManager->getShare($params['id']);
+
+		if ($shareInfo === false) {
+			return new \OC_OCS_Result(null, 404, 'Share does not exists');
```
s/exists/exist/

nickvergessen commented:
Maybe remote_shares/pending or do want to use the same endpoint for normal shares in the future?

This needs a PR to fix the URLs in the Pyoc client

rullzer commented:
> Maybe remote_shares/pending or do want to use the same endpoint for normal shares in the future?

Possible for sharing 2.0 yes.

DeepDiver1975 commented:
Hmmm ... do we really want to change the behavior of already existing public routes?

This is a no go from my perspective - but maybe I missed something.

rullzer commented:
> 

Hmmm ... do we really want to change the behavior of already existing public routes?
>
> This is a no go from my perspective - but maybe I missed something.

This route was introduced in 8.2. So it is not yet officially released.

DeepDiver1975 commented:
THX. Fine then

nickvergessen commented on apps/files_sharing/lib/external/manager.php:
```diff
@@ -180,9 +180,9 @@ private function setupMounts() {
+	public function getShare($id) {
 		$getShare = $this->connection->prepare('
-			SELECT `remote`, `remote_id`, `share_token`, `name`
+			SELECT *
```
do we want to expose all of the data?
Thinking about password and share_token, I'm not sure they should be leaked like that?

rullzer replied:
So I agree we should limit what we show. But we also show * in the listing of pending shares.. https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/external/manager.php#L400-L430

So this doesn't introduce anything new.

nickvergessen replied:
okay fine by me then

MorrisJobke commented:
@rullzer Ready to review or not?

rullzer commented:
@MorrisJobke not yet. I think I need to add some stuff. For #19501 to work properly.

nickvergessen commented:
btw while preparing a patch for the PyoC Client to change the URL, I realised we can't reuse the same endpoint for remote and normal shares! :bomb: 
The problem is that they have a different unique id. For remote shares we need to accept `oc_share_remote.id` for normal shares `oc_share.id` So they can not reuse the same endpoint.

rullzer commented:
@nickvergessen warg... unique id's are needed then indeed...

rullzer committed:
Made routes remote_shares only again

Since we need a unique id to accept/reject shares for now we keep the
pending shares under remote_shares.

* remote_shares/pending lists pending shares
* PUT/DELETE to remote_shares/pending/<ID> will accept/reject the share

rullzer commented:
@nickvergessen I guess this is then the most logical solution.

rullzer committed:
Extend share info

The data from the share_external is not to much. Thus we enrich this
data with info from the filecache.

This allows endpoints using this to actually show usefull information.

The filecache might not be up to date but that is a sacrifice we need to
make in terms of speed. Else the number of remote PROPFINDS grows
lineary with the number of remote shares wich will make this endpoint
practically unusable.

rullzer commented:
Ok extra info added.. do now you actually know something about the remote shares.

Review time.

CC: @nickvergessen @MorrisJobke @DeepDiver1975

rullzer committed:
Do not return unneeded info

rullzer commented:
Ok cleaned up a bit more so we return a little less cluttered stuff.

rullzer commented:
OCI failure seems unrelated.

nickvergessen commented:
Didn't test, but looks good.

PVince81 commented:
Works :+1: 

I get the JS code will need to be adjusted later.

nickvergessen commented:
PyOC patch https://github.com/owncloud/pyocclient/pull/122

rullzer commented:
> I get the JS code will need to be adjusted later.

@PVince81 Already done in #19514
But I did not want to have 1 huge PR.

PVince81 commented:
Ah yes, the JS code update is here https://github.com/owncloud/core/pull/19514

carlaschroder commented:
@rullzer Is this only for 8.2?

nickvergessen commented:
yes

carlaschroder commented:
thanks @nickvergessen
